Make it possible to move messages by recency

chdinesh1089 opened the pull request:
Fixes #20502

chdinesh1089 committed:
get_messages_backend: Add first_after_date anchor.

chdinesh1089 committed:
test_openapi: Extend schema_type() to consider format: date-time.

chdinesh1089 commented on zerver/tests/test_openapi.py:
```diff
@@ -815,6 +818,7 @@ def test_generate_and_render_curl_with_array_example(self) -> None:
             '    --data-urlencode \'narrow=[{"operand": "Denmark", "operator": "stream"}]\' \\',
             "    --data-urlencode client_gravatar=false \\",
             "    --data-urlencode apply_markdown=false \\",
+            "    --data-urlencode target_date=2022-03-17T18:37:30.806Z \\",
```
Is it correct to add this here for the sake of test though this doesn't match with `anchor=43`? Alternatively, we can add an `exclude` in `self.curl_example()`.

chdinesh1089 committed:
topics: Make it possible to move messages by recency.

Fixes #20502

chdinesh1089 committed:
move_topic_modal: Remove choose stream/topic help text.

This doesn't add much value as discussed in
https://chat.zulip.org/#narrow/stream/101-design/topic/Move.20messages.20by.20recency.20.2320502/near/1355873

alya commented:
3. Regardless of the answer to the above, let's just show the time down to the minute, without the seconds. I don't think that level of precision is helpful here.

alya commented:
4. I think we should drop either "last 7 days" or "last 10 days" as an option; it would be very hard to meaningfully pick between the two in this UI. Let's drop "last 7 days", I guess.

alya commented:
5. After seeing it, I think we need to rearrange the modal a bit. In particular:

a) Make the heading be: "Move messages from `<topic name>`.
b) Make the label above the destination just say "Destination:"
c) Make the label for which message to move say: "Move messages since:"
d) Accordingly, change the dropdown labels "Last X hours" -> "X hours ago"; the other labels stay the same.
e) Change the three-dot menu option label "Move topic" -> "Move messages".

alya commented:
1. "Error: No messages to move." sounds good to me.
2. Would it be possible to change it to work the other way, i.e. from when the dropdown selection is made? That way we can clearly communicate to the user what's actually going to happen.

`5`. Yes, exactly, I was thinking we would change the top header. Putting the name fo the topic in it seems pretty consistent with how we do other modals, e.g.:

![Screen Shot 2022-04-12 at 8 49 35 AM](https://user-images.githubusercontent.com/2090066/163002826-19205613-9e7f-4051-917b-f4a65306821f.png)
![Screen Shot 2022-04-12 at 8 49 15 AM](https://user-images.githubusercontent.com/2090066/163002841-68d1560e-737c-4ed7-b117-bc8dad29b86e.png)

If the topic name is too long to fit, I think we should just abbreviate it with a "...". Anyway, I think we should give it a try.

Thanks!

alya commented:
Hey @chdinesh1089 , this isn't as important as pushing your other PRs forward, but just a bump here in case you get a chance to address the feedback above.

asah commented on zerver/tests/test_message_fetch.py:
```diff
@@ -2787,6 +2787,129 @@ def messages_matches_ids(messages: List[Dict[str, Any]], message_ids: List[int])
+        self.assertEqual(result["messages"][0]["id"], message_ids[0])
+        self.assertEqual(result["anchor"], message_ids[0])
+
+    def test_first_after_date_anchor_different_narrow_first_msg(self) -> None:
```
duplicates code from test_first_after_date_anchor_target_date ?  refactor into a common setup function ?

chdinesh1089 replied:
I think little duplication is OK in tests. I feel like we'd be complicating it trying to refactor.

asah commented on zerver/tests/test_message_fetch.py:
```diff
@@ -2787,6 +2787,129 @@ def messages_matches_ids(messages: List[Dict[str, Any]], message_ids: List[int])
+        for i in range(3):
+            message_ids.append(self.send_stream_message(aaron, "Scotland", topic_name=topic))
+
+        # Edit timestamps to ensure they aren't same second
```
nice test.

chdinesh1089 replied:
Thanks! :blush:

asah commented on zerver/views/message_fetch.py:
```diff
@@ -1049,6 +1099,13 @@ def get_messages_backend(
+            anchor = find_first_after_date_anchor(sa_conn, user_profile, realm, narrow, target_date)
+
+            if anchor == -1:
+                raise NoMessageFoundAfterDate(target_date)
```
sorry if this is a dumb question (perhaps add a comment?) - why is no-message-found an error/exception and not just a normal behavior, i.e. returns zero messages?

put another way, I would have expected the API to return <up to n> messages between two time ranges, and zero messages to be an normal result.

chdinesh1089 replied:
We were thinking specifically in terms of this feature of moving messages by recency when writing this. See messages around https://chat.zulip.org/#narrow/stream/378-api-design/topic/GET.20.2Fmessages.2F.7Bmsg_id.7D.20anchor.3Dfirst_after_date/near/1354215 for previous discussion

Feel free to add your thoughts there too, to encourage some discussion over this.

timabbott commented:
@chdinesh1089 do you have time to rebase this? I'd like to push on integrating it.

chdinesh1089 commented:
Sure, will rebase during the weekend.

chdinesh1089 committed:
move_topic: Change wording to match with the option to move messages selectively.

Also includes some styling changes and tagged some strings for translation.

These are based on suggestions from @alya at
https://github.com/zulip/zulip/pull/21394#issuecomment-1090632442

chdinesh1089 commented:
@alya based on your above comments changed the UI to the following
![image](https://user-images.githubusercontent.com/44665669/181904572-ae14a03d-2d69-4446-a0e5-d1626086faa5.png)

@timabbott I've rebased this. We can merge the backend parts of this if it looks good.

Thanks for the review, @asah!

alya commented:
Thank you for trying the topic name in the header! Looking at your screenshot, the space looks a bit tight, and it might be better to include the stream name too, which would make it even tighter. Maybe we should try something like:

## Move messages
Moving messages from #stream > topic.
Destination:

alya commented:
It also looks like there's a failing test -- I don't know if that's expected?

alya commented:
Thanks for the work here, @chdinesh1089! Closing as I think we should implement #23741 instead.
